// @dynopii/callchimp/src/models/CampaignResponse.ts
/* tslint:disable */
/* eslint-disable */
/**
 * Callchimp Public API
 * ## Introduction👋 Introducing OpenAPI spec for doing almost anything in [callchimp.ai](https://callchimp.ai). CallChimp is a Gen AI Call Center Enhancing telecommunication with GPT-driven bulk calling. It is scalable, user-friendly, and customizable. Optimized for seamless integration and usability. ## API Categories📋 The APIs are divided in 9 categories listed below:   - Campaigns   - Supervisors   - Lists   - Subscribers   - Calls   - Phone Numbers   - Webhooks   - Scripts   - Voices  ## API Operations✅ ### Campaign Operations    - List Campaigns   - Create a Campaign   - Get Campaign by Id   - Delete Campaign by Id   - Update Campaign by Id   - Add Supervisors to Campaign by Id   - Remove Supervisors from Campaign by Id   - Upload audio file to Campaign by Id   - Search Campaign by Name or Id  ### Supervisor Operations    - List Supervisors   - Create a Supervisor   - Get Supervisor by Id   - Delete Supervisor by Id   - Update Supervisor by Id   - Send OTP to Supervisor by Id   - Verify Supervisor OTP by Id  ### List Operations    - List Lists   - Create a List   - Get List by Id   - Delete List by Id   - Update List by Id   - Search List by Name or Id   - Create Campaign with Default List  ### Subscriber Operations    - List Subscribers   - Create one or more Subscriber(s)   - Get Subscriber by Id   - Delete Subscriber by Id   - Update Subscriber by Id   - Get Subscriber by Vendor Lead Code   - Delete Subscriber by Vendor Lead Code   - Update Subscriber by Vendor Lead Code  ### Call Operations    - List Calls   - Create a Call   - Get Call by Id   - List Supervisor Inbound Calls   - Generate Call Reports  ### Phone Number Operations    - List Phone Numbers  ### Webhook Operations    - List Webhooks   - Create a Webhook   - Get Webhook by Id   - Delete Webhook by Id   - Update Webhook by Id  ### Script Operations    - List Scripts   - Create a Script   - Get Script by Id   - Delete Script by Id   - Update Script by Id  ### Voice Operations   - List Available Voices  ## Authentication🔐 Callchimp public API offers authentication with API Keys. All endpoints accepts a header named `x-api-key`. To get started follow the below steps:    1. Login to callchimp dashboard.   2. Click on your profile on the top-right corner.   3. Click on Settings.   4. On the settings page, click on `API Keys` tab.   5. Click on `Create` button, add a name and an expiry date and click on `Add`.   6. An API Key will be generated, please save the key somewhere safe as it won\'t be shown again!   7. You can use the API Key directly here in this playground to test out the APIs. 
 *
 * The version of the OpenAPI document: 1.0.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { mapValues } from '../runtime';
/**
 * 
 * @export
 * @interface CampaignResponse
 */
export interface CampaignResponse {
    /**
     * 
     * @type {Date}
     * @memberof CampaignResponse
     */
    addedAt?: Date;
    /**
     * 
     * @type {string}
     * @memberof CampaignResponse
     */
    botLanguage?: string;
    /**
     * Script for AI campaigns
     * @type {string}
     * @memberof CampaignResponse
     */
    chatScript?: string;
    /**
     * Expected language of the leads
     * @type {string}
     * @memberof CampaignResponse
     */
    customerLanguage?: string;
    /**
     * 
     * @type {number}
     * @memberof CampaignResponse
     */
    id?: number;
    /**
     * 
     * @type {boolean}
     * @memberof CampaignResponse
     */
    isActive?: boolean;
    /**
     * If active for AI campaigns, records all calls
     * @type {boolean}
     * @memberof CampaignResponse
     */
    isRecording?: boolean;
    /**
     * How many times call should be retried if not picked up
     * @type {number}
     * @memberof CampaignResponse
     */
    maxRetry?: CampaignResponseMaxRetryEnum;
    /**
     * 
     * @type {string}
     * @memberof CampaignResponse
     */
    name?: string;
    /**
     * Organization foreign key
     * @type {number}
     * @memberof CampaignResponse
     */
    organization?: number;
    /**
     * PhoneNumber foreign key
     * @type {number}
     * @memberof CampaignResponse
     */
    phoneNumber?: number;
    /**
     * List of supervisor foreign keys
     * @type {Array<number>}
     * @memberof CampaignResponse
     */
    supervisors?: Array<number>;
    /**
     * Type of campaign
     * @type {string}
     * @memberof CampaignResponse
     */
    type?: CampaignResponseTypeEnum;
}


/**
 * @export
 */
export const CampaignResponseMaxRetryEnum = {
    NUMBER_0: 0,
    NUMBER_1: 1,
    NUMBER_2: 2
} as const;
export type CampaignResponseMaxRetryEnum = typeof CampaignResponseMaxRetryEnum[keyof typeof CampaignResponseMaxRetryEnum];

/**
 * @export
 */
export const CampaignResponseTypeEnum = {
    Blastout: 'blastout',
    Outbound: 'outbound',
    Transactional: 'transactional'
} as const;
export type CampaignResponseTypeEnum = typeof CampaignResponseTypeEnum[keyof typeof CampaignResponseTypeEnum];


/**
 * Check if a given object implements the CampaignResponse interface.
 */
export function instanceOfCampaignResponse(value: object): value is CampaignResponse {
    return true;
}

export function CampaignResponseFromJSON(json: any): CampaignResponse {
    return CampaignResponseFromJSONTyped(json, false);
}

export function CampaignResponseFromJSONTyped(json: any, ignoreDiscriminator: boolean): CampaignResponse {
    if (json == null) {
        return json;
    }
    return {
        
        'addedAt': json['added_at'] == null ? undefined : (new Date(json['added_at'])),
        'botLanguage': json['bot_language'] == null ? undefined : json['bot_language'],
        'chatScript': json['chat_script'] == null ? undefined : json['chat_script'],
        'customerLanguage': json['customer_language'] == null ? undefined : json['customer_language'],
        'id': json['id'] == null ? undefined : json['id'],
        'isActive': json['is_active'] == null ? undefined : json['is_active'],
        'isRecording': json['is_recording'] == null ? undefined : json['is_recording'],
        'maxRetry': json['max_retry'] == null ? undefined : json['max_retry'],
        'name': json['name'] == null ? undefined : json['name'],
        'organization': json['organization'] == null ? undefined : json['organization'],
        'phoneNumber': json['phone_number'] == null ? undefined : json['phone_number'],
        'supervisors': json['supervisors'] == null ? undefined : json['supervisors'],
        'type': json['type'] == null ? undefined : json['type'],
    };
}

export function CampaignResponseToJSON(value?: CampaignResponse | null): any {
    if (value == null) {
        return value;
    }
    return {
        
        'added_at': value['addedAt'] == null ? undefined : ((value['addedAt']).toISOString()),
        'bot_language': value['botLanguage'],
        'chat_script': value['chatScript'],
        'customer_language': value['customerLanguage'],
        'id': value['id'],
        'is_active': value['isActive'],
        'is_recording': value['isRecording'],
        'max_retry': value['maxRetry'],
        'name': value['name'],
        'organization': value['organization'],
        'phone_number': value['phoneNumber'],
        'supervisors': value['supervisors'],
        'type': value['type'],
    };
}
