// @dynopii/callchimp/src/models/CampaignUpdateRequest.ts
/* tslint:disable */
/* eslint-disable */
/**
 * Callchimp Public API
 * ## Introduction👋 Introducing OpenAPI spec for doing almost anything in [callchimp.ai](https://callchimp.ai). CallChimp is a Gen AI Call Center Enhancing telecommunication with GPT-driven bulk calling. It is scalable, user-friendly, and customizable. Optimized for seamless integration and usability. ## API Categories📋 The APIs are divided in 9 categories listed below:   - Campaigns   - Supervisors   - Lists   - Subscribers   - Calls   - Phone Numbers   - Webhooks   - Scripts   - Voices  ## API Operations✅ ### Campaign Operations    - List Campaigns   - Create a Campaign   - Get Campaign by Id   - Delete Campaign by Id   - Update Campaign by Id   - Add Supervisors to Campaign by Id   - Remove Supervisors from Campaign by Id   - Upload audio file to Campaign by Id   - Search Campaign by Name or Id  ### Supervisor Operations    - List Supervisors   - Create a Supervisor   - Get Supervisor by Id   - Delete Supervisor by Id   - Update Supervisor by Id   - Send OTP to Supervisor by Id   - Verify Supervisor OTP by Id  ### List Operations    - List Lists   - Create a List   - Get List by Id   - Delete List by Id   - Update List by Id   - Search List by Name or Id   - Create Campaign with Default List  ### Subscriber Operations    - List Subscribers   - Create one or more Subscriber(s)   - Get Subscriber by Id   - Delete Subscriber by Id   - Update Subscriber by Id   - Get Subscriber by Vendor Lead Code   - Delete Subscriber by Vendor Lead Code   - Update Subscriber by Vendor Lead Code  ### Call Operations    - List Calls   - Create a Call   - Get Call by Id   - List Supervisor Inbound Calls   - Generate Call Reports  ### Phone Number Operations    - List Phone Numbers  ### Webhook Operations    - List Webhooks   - Create a Webhook   - Get Webhook by Id   - Delete Webhook by Id   - Update Webhook by Id  ### Script Operations    - List Scripts   - Create a Script   - Get Script by Id   - Delete Script by Id   - Update Script by Id  ### Voice Operations   - List Available Voices  ## Authentication🔐 Callchimp public API offers authentication with API Keys. All endpoints accepts a header named `x-api-key`. To get started follow the below steps:    1. Login to callchimp dashboard.   2. Click on your profile on the top-right corner.   3. Click on Settings.   4. On the settings page, click on `API Keys` tab.   5. Click on `Create` button, add a name and an expiry date and click on `Add`.   6. An API Key will be generated, please save the key somewhere safe as it won\'t be shown again!   7. You can use the API Key directly here in this playground to test out the APIs. 
 *
 * The version of the OpenAPI document: 1.0.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { mapValues } from '../runtime';
/**
 * 
 * @export
 * @interface CampaignUpdateRequest
 */
export interface CampaignUpdateRequest {
    /**
     * 
     * @type {string}
     * @memberof CampaignUpdateRequest
     */
    name?: string;
    /**
     * How many times call should be retried if not picked up
     * @type {number}
     * @memberof CampaignUpdateRequest
     */
    maxRetry?: CampaignUpdateRequestMaxRetryEnum;
    /**
     * PhoneNumber foreign key
     * @type {number}
     * @memberof CampaignUpdateRequest
     */
    phoneNumber?: number;
    /**
     * Type of campaign
     * @type {string}
     * @memberof CampaignUpdateRequest
     */
    type?: CampaignUpdateRequestTypeEnum;
    /**
     * Template text with variable placeholders, where variables are represented like `{var1}`. Only required for `transactional` campaigns.
     * @type {string}
     * @memberof CampaignUpdateRequest
     */
    transactionTemplate?: string;
    /**
     * Script id, Can be found from the Script Listing API. This is the GenAI prompt telling the bot how to behave when asked a question or how to continue the conversation.
     * @type {number}
     * @memberof CampaignUpdateRequest
     */
    script?: number;
    /**
     * Voice id. Can be found from the Voice Listing API. Head over to [Voice Garden](https://voices.callchimp.ai/) by Callchimp for samples.
     * @type {number}
     * @memberof CampaignUpdateRequest
     */
    voice?: number;
}


/**
 * @export
 */
export const CampaignUpdateRequestMaxRetryEnum = {
    NUMBER_0: 0,
    NUMBER_1: 1,
    NUMBER_2: 2
} as const;
export type CampaignUpdateRequestMaxRetryEnum = typeof CampaignUpdateRequestMaxRetryEnum[keyof typeof CampaignUpdateRequestMaxRetryEnum];

/**
 * @export
 */
export const CampaignUpdateRequestTypeEnum = {
    Blastout: 'blastout',
    Outbound: 'outbound',
    Transactional: 'transactional'
} as const;
export type CampaignUpdateRequestTypeEnum = typeof CampaignUpdateRequestTypeEnum[keyof typeof CampaignUpdateRequestTypeEnum];


/**
 * Check if a given object implements the CampaignUpdateRequest interface.
 */
export function instanceOfCampaignUpdateRequest(value: object): value is CampaignUpdateRequest {
    return true;
}

export function CampaignUpdateRequestFromJSON(json: any): CampaignUpdateRequest {
    return CampaignUpdateRequestFromJSONTyped(json, false);
}

export function CampaignUpdateRequestFromJSONTyped(json: any, ignoreDiscriminator: boolean): CampaignUpdateRequest {
    if (json == null) {
        return json;
    }
    return {
        
        'name': json['name'] == null ? undefined : json['name'],
        'maxRetry': json['max_retry'] == null ? undefined : json['max_retry'],
        'phoneNumber': json['phone_number'] == null ? undefined : json['phone_number'],
        'type': json['type'] == null ? undefined : json['type'],
        'transactionTemplate': json['transaction_template'] == null ? undefined : json['transaction_template'],
        'script': json['script'] == null ? undefined : json['script'],
        'voice': json['voice'] == null ? undefined : json['voice'],
    };
}

export function CampaignUpdateRequestToJSON(value?: CampaignUpdateRequest | null): any {
    if (value == null) {
        return value;
    }
    return {
        
        'name': value['name'],
        'max_retry': value['maxRetry'],
        'phone_number': value['phoneNumber'],
        'type': value['type'],
        'transaction_template': value['transactionTemplate'],
        'script': value['script'],
        'voice': value['voice'],
    };
}
