// @dynopii/callchimp/src/models/SubscriberResponse.ts
/* tslint:disable */
/* eslint-disable */
/**
 * Callchimp Public API
 * ## Introduction👋 Introducing OpenAPI spec for doing almost anything in [callchimp.ai](https://callchimp.ai). CallChimp is a Gen AI Call Center Enhancing telecommunication with GPT-driven bulk calling. It is scalable, user-friendly, and customizable. Optimized for seamless integration and usability. ## API Categories📋 The APIs are divided in 9 categories listed below:   - Campaigns   - Supervisors   - Lists   - Subscribers   - Calls   - Phone Numbers   - Webhooks   - Scripts   - Voices  ## API Operations✅ ### Campaign Operations    - List Campaigns   - Create a Campaign   - Get Campaign by Id   - Delete Campaign by Id   - Update Campaign by Id   - Add Supervisors to Campaign by Id   - Remove Supervisors from Campaign by Id   - Upload audio file to Campaign by Id   - Search Campaign by Name or Id  ### Supervisor Operations    - List Supervisors   - Create a Supervisor   - Get Supervisor by Id   - Delete Supervisor by Id   - Update Supervisor by Id   - Send OTP to Supervisor by Id   - Verify Supervisor OTP by Id  ### List Operations    - List Lists   - Create a List   - Get List by Id   - Delete List by Id   - Update List by Id   - Search List by Name or Id   - Create Campaign with Default List  ### Subscriber Operations    - List Subscribers   - Create one or more Subscriber(s)   - Get Subscriber by Id   - Delete Subscriber by Id   - Update Subscriber by Id   - Get Subscriber by Vendor Lead Code   - Delete Subscriber by Vendor Lead Code   - Update Subscriber by Vendor Lead Code  ### Call Operations    - List Calls   - Create a Call   - Get Call by Id   - List Supervisor Inbound Calls   - Generate Call Reports  ### Phone Number Operations    - List Phone Numbers  ### Webhook Operations    - List Webhooks   - Create a Webhook   - Get Webhook by Id   - Delete Webhook by Id   - Update Webhook by Id  ### Script Operations    - List Scripts   - Create a Script   - Get Script by Id   - Delete Script by Id   - Update Script by Id  ### Voice Operations   - List Available Voices  ## Authentication🔐 Callchimp public API offers authentication with API Keys. All endpoints accepts a header named `x-api-key`. To get started follow the below steps:    1. Login to callchimp dashboard.   2. Click on your profile on the top-right corner.   3. Click on Settings.   4. On the settings page, click on `API Keys` tab.   5. Click on `Create` button, add a name and an expiry date and click on `Add`.   6. An API Key will be generated, please save the key somewhere safe as it won\'t be shown again!   7. You can use the API Key directly here in this playground to test out the APIs. 
 *
 * The version of the OpenAPI document: 1.0.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { mapValues } from '../runtime';
/**
 * 
 * @export
 * @interface SubscriberResponse
 */
export interface SubscriberResponse {
    /**
     * 
     * @type {number}
     * @memberof SubscriberResponse
     */
    id?: number;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    status?: string;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    user?: string;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    vendorLeadCode?: string;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    sourceId?: string;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    phoneCode?: string;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    phoneNumber?: string;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    title?: string;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    firstName?: string;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    middleInitial?: string;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    lastName?: string;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    address1?: string;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    address2?: string;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    address3?: string;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    city?: string;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    state?: string;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    province?: string;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    postalCode?: string;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    countryCode?: string;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    gender?: string;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    dateOfBirth?: string;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    altPhone?: string;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    email?: string;
    /**
     * 
     * @type {number}
     * @memberof SubscriberResponse
     */
    calledCount?: number;
    /**
     * 
     * @type {string}
     * @memberof SubscriberResponse
     */
    rank?: string;
    /**
     * Leadlist foreign key
     * @type {number}
     * @memberof SubscriberResponse
     */
    leadlist?: number;
    /**
     * Organization foreign key
     * @type {number}
     * @memberof SubscriberResponse
     */
    organization?: number;
}

/**
 * Check if a given object implements the SubscriberResponse interface.
 */
export function instanceOfSubscriberResponse(value: object): value is SubscriberResponse {
    return true;
}

export function SubscriberResponseFromJSON(json: any): SubscriberResponse {
    return SubscriberResponseFromJSONTyped(json, false);
}

export function SubscriberResponseFromJSONTyped(json: any, ignoreDiscriminator: boolean): SubscriberResponse {
    if (json == null) {
        return json;
    }
    return {
        
        'id': json['id'] == null ? undefined : json['id'],
        'status': json['status'] == null ? undefined : json['status'],
        'user': json['user'] == null ? undefined : json['user'],
        'vendorLeadCode': json['vendor_lead_code'] == null ? undefined : json['vendor_lead_code'],
        'sourceId': json['source_id'] == null ? undefined : json['source_id'],
        'phoneCode': json['phone_code'] == null ? undefined : json['phone_code'],
        'phoneNumber': json['phone_number'] == null ? undefined : json['phone_number'],
        'title': json['title'] == null ? undefined : json['title'],
        'firstName': json['first_name'] == null ? undefined : json['first_name'],
        'middleInitial': json['middle_initial'] == null ? undefined : json['middle_initial'],
        'lastName': json['last_name'] == null ? undefined : json['last_name'],
        'address1': json['address1'] == null ? undefined : json['address1'],
        'address2': json['address2'] == null ? undefined : json['address2'],
        'address3': json['address3'] == null ? undefined : json['address3'],
        'city': json['city'] == null ? undefined : json['city'],
        'state': json['state'] == null ? undefined : json['state'],
        'province': json['province'] == null ? undefined : json['province'],
        'postalCode': json['postal_code'] == null ? undefined : json['postal_code'],
        'countryCode': json['country_code'] == null ? undefined : json['country_code'],
        'gender': json['gender'] == null ? undefined : json['gender'],
        'dateOfBirth': json['date_of_birth'] == null ? undefined : json['date_of_birth'],
        'altPhone': json['alt_phone'] == null ? undefined : json['alt_phone'],
        'email': json['email'] == null ? undefined : json['email'],
        'calledCount': json['called_count'] == null ? undefined : json['called_count'],
        'rank': json['rank'] == null ? undefined : json['rank'],
        'leadlist': json['leadlist'] == null ? undefined : json['leadlist'],
        'organization': json['organization'] == null ? undefined : json['organization'],
    };
}

export function SubscriberResponseToJSON(value?: SubscriberResponse | null): any {
    if (value == null) {
        return value;
    }
    return {
        
        'id': value['id'],
        'status': value['status'],
        'user': value['user'],
        'vendor_lead_code': value['vendorLeadCode'],
        'source_id': value['sourceId'],
        'phone_code': value['phoneCode'],
        'phone_number': value['phoneNumber'],
        'title': value['title'],
        'first_name': value['firstName'],
        'middle_initial': value['middleInitial'],
        'last_name': value['lastName'],
        'address1': value['address1'],
        'address2': value['address2'],
        'address3': value['address3'],
        'city': value['city'],
        'state': value['state'],
        'province': value['province'],
        'postal_code': value['postalCode'],
        'country_code': value['countryCode'],
        'gender': value['gender'],
        'date_of_birth': value['dateOfBirth'],
        'alt_phone': value['altPhone'],
        'email': value['email'],
        'called_count': value['calledCount'],
        'rank': value['rank'],
        'leadlist': value['leadlist'],
        'organization': value['organization'],
    };
}
