// @dynopii/callchimp/src/apis/CallsApi.ts
/* tslint:disable */
/* eslint-disable */
/**
 * Callchimp Public API
 * ## Introduction👋 Introducing OpenAPI spec for doing almost anything in [callchimp.ai](https://callchimp.ai). CallChimp is a Gen AI Call Center Enhancing telecommunication with GPT-driven bulk calling. It is scalable, user-friendly, and customizable. Optimized for seamless integration and usability. ## API Categories📋 The APIs are divided in 9 categories listed below:   - Campaigns   - Supervisors   - Lists   - Subscribers   - Calls   - Phone Numbers   - Webhooks   - Scripts   - Voices  ## API Operations✅ ### Campaign Operations    - List Campaigns   - Create a Campaign   - Get Campaign by Id   - Delete Campaign by Id   - Update Campaign by Id   - Add Supervisors to Campaign by Id   - Remove Supervisors from Campaign by Id   - Upload audio file to Campaign by Id   - Search Campaign by Name or Id  ### Supervisor Operations    - List Supervisors   - Create a Supervisor   - Get Supervisor by Id   - Delete Supervisor by Id   - Update Supervisor by Id   - Send OTP to Supervisor by Id   - Verify Supervisor OTP by Id  ### List Operations    - List Lists   - Create a List   - Get List by Id   - Delete List by Id   - Update List by Id   - Search List by Name or Id   - Create Campaign with Default List  ### Subscriber Operations    - List Subscribers   - Create one or more Subscriber(s)   - Get Subscriber by Id   - Delete Subscriber by Id   - Update Subscriber by Id   - Get Subscriber by Vendor Lead Code   - Delete Subscriber by Vendor Lead Code   - Update Subscriber by Vendor Lead Code  ### Call Operations    - List Calls   - Create a Call   - Get Call by Id   - List Supervisor Inbound Calls   - Generate Call Reports  ### Phone Number Operations    - List Phone Numbers  ### Webhook Operations    - List Webhooks   - Create a Webhook   - Get Webhook by Id   - Delete Webhook by Id   - Update Webhook by Id  ### Script Operations    - List Scripts   - Create a Script   - Get Script by Id   - Delete Script by Id   - Update Script by Id  ### Voice Operations   - List Available Voices  ## Authentication🔐 Callchimp public API offers authentication with API Keys. All endpoints accepts a header named `x-api-key`. To get started follow the below steps:    1. Login to callchimp dashboard.   2. Click on your profile on the top-right corner.   3. Click on Settings.   4. On the settings page, click on `API Keys` tab.   5. Click on `Create` button, add a name and an expiry date and click on `Add`.   6. An API Key will be generated, please save the key somewhere safe as it won\'t be shown again!   7. You can use the API Key directly here in this playground to test out the APIs. 
 *
 * The version of the OpenAPI document: 1.0.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


import * as runtime from '../runtime';
import type {
  CallListResponse,
  CallReportRequest,
  CallReportResponse,
  CallResponse,
  CallsPostRequest,
  InboundCallListResponse,
  Model4XXResponse,
} from '../models/index';
import {
    CallListResponseFromJSON,
    CallListResponseToJSON,
    CallReportRequestFromJSON,
    CallReportRequestToJSON,
    CallReportResponseFromJSON,
    CallReportResponseToJSON,
    CallResponseFromJSON,
    CallResponseToJSON,
    CallsPostRequestFromJSON,
    CallsPostRequestToJSON,
    InboundCallListResponseFromJSON,
    InboundCallListResponseToJSON,
    Model4XXResponseFromJSON,
    Model4XXResponseToJSON,
} from '../models/index';

export interface CallsGetRequest {
    id: number;
}

export interface CallsListInboundRequest {
    page?: number;
}

export interface CallsListOutboundRequest {
    page?: number;
}

export interface CallsPostOperationRequest {
    callsPostRequest: CallsPostRequest;
}

export interface CallsReportsRequest {
    callReportRequest: CallReportRequest;
    retries?: number;
}

/**
 * 
 */
export class CallsApi extends runtime.BaseAPI {

    /**
     * 
     * Get Call by Id
     */
    async callsGetRaw(requestParameters: CallsGetRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<CallResponse>> {
        if (requestParameters['id'] == null) {
            throw new runtime.RequiredError(
                'id',
                'Required parameter "id" was null or undefined when calling callsGet().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/calls/{Id}`.replace(`{${"Id"}}`, encodeURIComponent(String(requestParameters['id']))),
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => CallResponseFromJSON(jsonValue));
    }

    /**
     * 
     * Get Call by Id
     */
    async callsGet(requestParameters: CallsGetRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<CallResponse> {
        const response = await this.callsGetRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * List Supervisor Inbound Calls
     */
    async callsListInboundRaw(requestParameters: CallsListInboundRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<InboundCallListResponse>> {
        const queryParameters: any = {};

        if (requestParameters['page'] != null) {
            queryParameters['page'] = requestParameters['page'];
        }

        const headerParameters: runtime.HTTPHeaders = {};

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/calls/inbound`,
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => InboundCallListResponseFromJSON(jsonValue));
    }

    /**
     * 
     * List Supervisor Inbound Calls
     */
    async callsListInbound(requestParameters: CallsListInboundRequest = {}, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<InboundCallListResponse> {
        const response = await this.callsListInboundRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * List Calls
     */
    async callsListOutboundRaw(requestParameters: CallsListOutboundRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<CallListResponse>> {
        const queryParameters: any = {};

        if (requestParameters['page'] != null) {
            queryParameters['page'] = requestParameters['page'];
        }

        const headerParameters: runtime.HTTPHeaders = {};

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/calls`,
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => CallListResponseFromJSON(jsonValue));
    }

    /**
     * 
     * List Calls
     */
    async callsListOutbound(requestParameters: CallsListOutboundRequest = {}, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<CallListResponse> {
        const response = await this.callsListOutboundRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * Create a Call
     */
    async callsPostRaw(requestParameters: CallsPostOperationRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<CallResponse>> {
        if (requestParameters['callsPostRequest'] == null) {
            throw new runtime.RequiredError(
                'callsPostRequest',
                'Required parameter "callsPostRequest" was null or undefined when calling callsPost().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        headerParameters['Content-Type'] = 'application/json';

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/calls`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: CallsPostRequestToJSON(requestParameters['callsPostRequest']),
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => CallResponseFromJSON(jsonValue));
    }

    /**
     * 
     * Create a Call
     */
    async callsPost(requestParameters: CallsPostOperationRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<CallResponse> {
        const response = await this.callsPostRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * Generate Call Reports
     */
    async callsReportsRaw(requestParameters: CallsReportsRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<CallReportResponse>> {
        if (requestParameters['callReportRequest'] == null) {
            throw new runtime.RequiredError(
                'callReportRequest',
                'Required parameter "callReportRequest" was null or undefined when calling callsReports().'
            );
        }

        const queryParameters: any = {};

        if (requestParameters['retries'] != null) {
            queryParameters['retries'] = requestParameters['retries'];
        }

        const headerParameters: runtime.HTTPHeaders = {};

        headerParameters['Content-Type'] = 'application/json';

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/calls/reports`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: CallReportRequestToJSON(requestParameters['callReportRequest']),
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => CallReportResponseFromJSON(jsonValue));
    }

    /**
     * 
     * Generate Call Reports
     */
    async callsReports(requestParameters: CallsReportsRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<CallReportResponse> {
        const response = await this.callsReportsRaw(requestParameters, initOverrides);
        return await response.value();
    }

}
