// @dynopii/callchimp/src/apis/CampaignsApi.ts
/* tslint:disable */
/* eslint-disable */
/**
 * Callchimp Public API
 * ## Introduction👋 Introducing OpenAPI spec for doing almost anything in [callchimp.ai](https://callchimp.ai). CallChimp is a Gen AI Call Center Enhancing telecommunication with GPT-driven bulk calling. It is scalable, user-friendly, and customizable. Optimized for seamless integration and usability. ## API Categories📋 The APIs are divided in 9 categories listed below:   - Campaigns   - Supervisors   - Lists   - Subscribers   - Calls   - Phone Numbers   - Webhooks   - Scripts   - Voices  ## API Operations✅ ### Campaign Operations    - List Campaigns   - Create a Campaign   - Get Campaign by Id   - Delete Campaign by Id   - Update Campaign by Id   - Add Supervisors to Campaign by Id   - Remove Supervisors from Campaign by Id   - Upload audio file to Campaign by Id   - Search Campaign by Name or Id  ### Supervisor Operations    - List Supervisors   - Create a Supervisor   - Get Supervisor by Id   - Delete Supervisor by Id   - Update Supervisor by Id   - Send OTP to Supervisor by Id   - Verify Supervisor OTP by Id  ### List Operations    - List Lists   - Create a List   - Get List by Id   - Delete List by Id   - Update List by Id   - Search List by Name or Id   - Create Campaign with Default List  ### Subscriber Operations    - List Subscribers   - Create one or more Subscriber(s)   - Get Subscriber by Id   - Delete Subscriber by Id   - Update Subscriber by Id   - Get Subscriber by Vendor Lead Code   - Delete Subscriber by Vendor Lead Code   - Update Subscriber by Vendor Lead Code  ### Call Operations    - List Calls   - Create a Call   - Get Call by Id   - List Supervisor Inbound Calls   - Generate Call Reports  ### Phone Number Operations    - List Phone Numbers  ### Webhook Operations    - List Webhooks   - Create a Webhook   - Get Webhook by Id   - Delete Webhook by Id   - Update Webhook by Id  ### Script Operations    - List Scripts   - Create a Script   - Get Script by Id   - Delete Script by Id   - Update Script by Id  ### Voice Operations   - List Available Voices  ## Authentication🔐 Callchimp public API offers authentication with API Keys. All endpoints accepts a header named `x-api-key`. To get started follow the below steps:    1. Login to callchimp dashboard.   2. Click on your profile on the top-right corner.   3. Click on Settings.   4. On the settings page, click on `API Keys` tab.   5. Click on `Create` button, add a name and an expiry date and click on `Add`.   6. An API Key will be generated, please save the key somewhere safe as it won\'t be shown again!   7. You can use the API Key directly here in this playground to test out the APIs. 
 *
 * The version of the OpenAPI document: 1.0.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


import * as runtime from '../runtime';
import type {
  CampaignAddSuperRequest,
  CampaignAddSuperResponse,
  CampaignFindRequest,
  CampaignListResponse,
  CampaignRemoveSuperRequest,
  CampaignRemoveSuperResponse,
  CampaignRequest,
  CampaignResponse,
  CampaignUpdateRequest,
  CampaignUploadAudioResponse,
  Model4XXResponse,
} from '../models/index';
import {
    CampaignAddSuperRequestFromJSON,
    CampaignAddSuperRequestToJSON,
    CampaignAddSuperResponseFromJSON,
    CampaignAddSuperResponseToJSON,
    CampaignFindRequestFromJSON,
    CampaignFindRequestToJSON,
    CampaignListResponseFromJSON,
    CampaignListResponseToJSON,
    CampaignRemoveSuperRequestFromJSON,
    CampaignRemoveSuperRequestToJSON,
    CampaignRemoveSuperResponseFromJSON,
    CampaignRemoveSuperResponseToJSON,
    CampaignRequestFromJSON,
    CampaignRequestToJSON,
    CampaignResponseFromJSON,
    CampaignResponseToJSON,
    CampaignUpdateRequestFromJSON,
    CampaignUpdateRequestToJSON,
    CampaignUploadAudioResponseFromJSON,
    CampaignUploadAudioResponseToJSON,
    Model4XXResponseFromJSON,
    Model4XXResponseToJSON,
} from '../models/index';

export interface CampaignsAddsuperRequest {
    id: number;
    campaignAddSuperRequest: CampaignAddSuperRequest;
}

export interface CampaignsDeleteRequest {
    id: number;
}

export interface CampaignsFindRequest {
    campaignFindRequest: CampaignFindRequest;
    page?: number;
}

export interface CampaignsGetRequest {
    id: number;
}

export interface CampaignsListRequest {
    page?: number;
}

export interface CampaignsPostRequest {
    campaignRequest: CampaignRequest;
}

export interface CampaignsRemovesuperRequest {
    id: number;
    campaignRemoveSuperRequest: CampaignRemoveSuperRequest;
}

export interface CampaignsUpdateRequest {
    id: number;
    campaignUpdateRequest: CampaignUpdateRequest;
}

export interface CampaignsUploadblastRequest {
    id: number;
    file: Blob;
}

/**
 * 
 */
export class CampaignsApi extends runtime.BaseAPI {

    /**
     * 
     * Add Supervisors to Campaign by Id
     */
    async campaignsAddsuperRaw(requestParameters: CampaignsAddsuperRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<CampaignAddSuperResponse>> {
        if (requestParameters['id'] == null) {
            throw new runtime.RequiredError(
                'id',
                'Required parameter "id" was null or undefined when calling campaignsAddsuper().'
            );
        }

        if (requestParameters['campaignAddSuperRequest'] == null) {
            throw new runtime.RequiredError(
                'campaignAddSuperRequest',
                'Required parameter "campaignAddSuperRequest" was null or undefined when calling campaignsAddsuper().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        headerParameters['Content-Type'] = 'application/json';

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/campaigns/{Id}/add_super`.replace(`{${"Id"}}`, encodeURIComponent(String(requestParameters['id']))),
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: CampaignAddSuperRequestToJSON(requestParameters['campaignAddSuperRequest']),
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => CampaignAddSuperResponseFromJSON(jsonValue));
    }

    /**
     * 
     * Add Supervisors to Campaign by Id
     */
    async campaignsAddsuper(requestParameters: CampaignsAddsuperRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<CampaignAddSuperResponse> {
        const response = await this.campaignsAddsuperRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * Delete Campaign by Id
     */
    async campaignsDeleteRaw(requestParameters: CampaignsDeleteRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<void>> {
        if (requestParameters['id'] == null) {
            throw new runtime.RequiredError(
                'id',
                'Required parameter "id" was null or undefined when calling campaignsDelete().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/campaigns/{Id}`.replace(`{${"Id"}}`, encodeURIComponent(String(requestParameters['id']))),
            method: 'DELETE',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);

        return new runtime.VoidApiResponse(response);
    }

    /**
     * 
     * Delete Campaign by Id
     */
    async campaignsDelete(requestParameters: CampaignsDeleteRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<void> {
        await this.campaignsDeleteRaw(requestParameters, initOverrides);
    }

    /**
     * 
     * Search Campaign by Name or Id
     */
    async campaignsFindRaw(requestParameters: CampaignsFindRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<Array<CampaignResponse>>> {
        if (requestParameters['campaignFindRequest'] == null) {
            throw new runtime.RequiredError(
                'campaignFindRequest',
                'Required parameter "campaignFindRequest" was null or undefined when calling campaignsFind().'
            );
        }

        const queryParameters: any = {};

        if (requestParameters['page'] != null) {
            queryParameters['page'] = requestParameters['page'];
        }

        const headerParameters: runtime.HTTPHeaders = {};

        headerParameters['Content-Type'] = 'application/json';

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/campaigns/find`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: CampaignFindRequestToJSON(requestParameters['campaignFindRequest']),
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => jsonValue.map(CampaignResponseFromJSON));
    }

    /**
     * 
     * Search Campaign by Name or Id
     */
    async campaignsFind(requestParameters: CampaignsFindRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<Array<CampaignResponse>> {
        const response = await this.campaignsFindRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * Get Campaign by Id
     */
    async campaignsGetRaw(requestParameters: CampaignsGetRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<CampaignResponse>> {
        if (requestParameters['id'] == null) {
            throw new runtime.RequiredError(
                'id',
                'Required parameter "id" was null or undefined when calling campaignsGet().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/campaigns/{Id}`.replace(`{${"Id"}}`, encodeURIComponent(String(requestParameters['id']))),
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => CampaignResponseFromJSON(jsonValue));
    }

    /**
     * 
     * Get Campaign by Id
     */
    async campaignsGet(requestParameters: CampaignsGetRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<CampaignResponse> {
        const response = await this.campaignsGetRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * ##### A paginated list of campaigns 
     * List campaigns
     */
    async campaignsListRaw(requestParameters: CampaignsListRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<CampaignListResponse>> {
        const queryParameters: any = {};

        if (requestParameters['page'] != null) {
            queryParameters['page'] = requestParameters['page'];
        }

        const headerParameters: runtime.HTTPHeaders = {};

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/campaigns`,
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => CampaignListResponseFromJSON(jsonValue));
    }

    /**
     * ##### A paginated list of campaigns 
     * List campaigns
     */
    async campaignsList(requestParameters: CampaignsListRequest = {}, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<CampaignListResponse> {
        const response = await this.campaignsListRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * #####  With callchimp.ai, you have access to 4 different types of campaigns:  - **Outbound Bulk**: Campaign that play pre recorded audio to the subscribers / leads. One way broadcast campaign.  - **Outbound AI**: GPT driven outbound AI campaign that runs according to the given script. Bidirectional campaign.   - **Inbound AI**: GPT driven inbound AI campaign that runs according to the given script. Bidirectional campaign.  - **Transactional**: Campaign that play dynamic text to audio to the subscribers / leads. One way campaign. 
     * Create a Campaign
     */
    async campaignsPostRaw(requestParameters: CampaignsPostRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<CampaignResponse>> {
        if (requestParameters['campaignRequest'] == null) {
            throw new runtime.RequiredError(
                'campaignRequest',
                'Required parameter "campaignRequest" was null or undefined when calling campaignsPost().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        headerParameters['Content-Type'] = 'application/json';

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/campaigns`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: CampaignRequestToJSON(requestParameters['campaignRequest']),
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => CampaignResponseFromJSON(jsonValue));
    }

    /**
     * #####  With callchimp.ai, you have access to 4 different types of campaigns:  - **Outbound Bulk**: Campaign that play pre recorded audio to the subscribers / leads. One way broadcast campaign.  - **Outbound AI**: GPT driven outbound AI campaign that runs according to the given script. Bidirectional campaign.   - **Inbound AI**: GPT driven inbound AI campaign that runs according to the given script. Bidirectional campaign.  - **Transactional**: Campaign that play dynamic text to audio to the subscribers / leads. One way campaign. 
     * Create a Campaign
     */
    async campaignsPost(requestParameters: CampaignsPostRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<CampaignResponse> {
        const response = await this.campaignsPostRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * Remove Supervisors from Campaign by Id
     */
    async campaignsRemovesuperRaw(requestParameters: CampaignsRemovesuperRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<CampaignRemoveSuperResponse>> {
        if (requestParameters['id'] == null) {
            throw new runtime.RequiredError(
                'id',
                'Required parameter "id" was null or undefined when calling campaignsRemovesuper().'
            );
        }

        if (requestParameters['campaignRemoveSuperRequest'] == null) {
            throw new runtime.RequiredError(
                'campaignRemoveSuperRequest',
                'Required parameter "campaignRemoveSuperRequest" was null or undefined when calling campaignsRemovesuper().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        headerParameters['Content-Type'] = 'application/json';

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/campaigns/{Id}/remove_super`.replace(`{${"Id"}}`, encodeURIComponent(String(requestParameters['id']))),
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: CampaignRemoveSuperRequestToJSON(requestParameters['campaignRemoveSuperRequest']),
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => CampaignRemoveSuperResponseFromJSON(jsonValue));
    }

    /**
     * 
     * Remove Supervisors from Campaign by Id
     */
    async campaignsRemovesuper(requestParameters: CampaignsRemovesuperRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<CampaignRemoveSuperResponse> {
        const response = await this.campaignsRemovesuperRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * Update Campaign by Id
     */
    async campaignsUpdateRaw(requestParameters: CampaignsUpdateRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<CampaignResponse>> {
        if (requestParameters['id'] == null) {
            throw new runtime.RequiredError(
                'id',
                'Required parameter "id" was null or undefined when calling campaignsUpdate().'
            );
        }

        if (requestParameters['campaignUpdateRequest'] == null) {
            throw new runtime.RequiredError(
                'campaignUpdateRequest',
                'Required parameter "campaignUpdateRequest" was null or undefined when calling campaignsUpdate().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        headerParameters['Content-Type'] = 'application/json';

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/campaigns/{Id}`.replace(`{${"Id"}}`, encodeURIComponent(String(requestParameters['id']))),
            method: 'PATCH',
            headers: headerParameters,
            query: queryParameters,
            body: CampaignUpdateRequestToJSON(requestParameters['campaignUpdateRequest']),
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => CampaignResponseFromJSON(jsonValue));
    }

    /**
     * 
     * Update Campaign by Id
     */
    async campaignsUpdate(requestParameters: CampaignsUpdateRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<CampaignResponse> {
        const response = await this.campaignsUpdateRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * Upload audio file to Campaign by Id
     */
    async campaignsUploadblastRaw(requestParameters: CampaignsUploadblastRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<CampaignUploadAudioResponse>> {
        if (requestParameters['id'] == null) {
            throw new runtime.RequiredError(
                'id',
                'Required parameter "id" was null or undefined when calling campaignsUploadblast().'
            );
        }

        if (requestParameters['file'] == null) {
            throw new runtime.RequiredError(
                'file',
                'Required parameter "file" was null or undefined when calling campaignsUploadblast().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const consumes: runtime.Consume[] = [
            { contentType: 'multipart/form-data' },
        ];
        // @ts-ignore: canConsumeForm may be unused
        const canConsumeForm = runtime.canConsumeForm(consumes);

        let formParams: { append(param: string, value: any): any };
        let useForm = false;
        // use FormData to transmit files using content-type "multipart/form-data"
        useForm = canConsumeForm;
        if (useForm) {
            formParams = new FormData();
        } else {
            formParams = new URLSearchParams();
        }

        if (requestParameters['file'] != null) {
            formParams.append('file', requestParameters['file'] as any);
        }

        const response = await this.request({
            path: `/campaigns/{Id}/upload_blast_file`.replace(`{${"Id"}}`, encodeURIComponent(String(requestParameters['id']))),
            method: 'PATCH',
            headers: headerParameters,
            query: queryParameters,
            body: formParams,
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => CampaignUploadAudioResponseFromJSON(jsonValue));
    }

    /**
     * 
     * Upload audio file to Campaign by Id
     */
    async campaignsUploadblast(requestParameters: CampaignsUploadblastRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<CampaignUploadAudioResponse> {
        const response = await this.campaignsUploadblastRaw(requestParameters, initOverrides);
        return await response.value();
    }

}
