// @dynopii/callchimp/src/apis/ListsApi.ts
/* tslint:disable */
/* eslint-disable */
/**
 * Callchimp Public API
 * ## Introduction👋 Introducing OpenAPI spec for doing almost anything in [callchimp.ai](https://callchimp.ai). CallChimp is a Gen AI Call Center Enhancing telecommunication with GPT-driven bulk calling. It is scalable, user-friendly, and customizable. Optimized for seamless integration and usability. ## API Categories📋 The APIs are divided in 9 categories listed below:   - Campaigns   - Supervisors   - Lists   - Subscribers   - Calls   - Phone Numbers   - Webhooks   - Scripts   - Voices  ## API Operations✅ ### Campaign Operations    - List Campaigns   - Create a Campaign   - Get Campaign by Id   - Delete Campaign by Id   - Update Campaign by Id   - Add Supervisors to Campaign by Id   - Remove Supervisors from Campaign by Id   - Upload audio file to Campaign by Id   - Search Campaign by Name or Id  ### Supervisor Operations    - List Supervisors   - Create a Supervisor   - Get Supervisor by Id   - Delete Supervisor by Id   - Update Supervisor by Id   - Send OTP to Supervisor by Id   - Verify Supervisor OTP by Id  ### List Operations    - List Lists   - Create a List   - Get List by Id   - Delete List by Id   - Update List by Id   - Search List by Name or Id   - Create Campaign with Default List  ### Subscriber Operations    - List Subscribers   - Create one or more Subscriber(s)   - Get Subscriber by Id   - Delete Subscriber by Id   - Update Subscriber by Id   - Get Subscriber by Vendor Lead Code   - Delete Subscriber by Vendor Lead Code   - Update Subscriber by Vendor Lead Code  ### Call Operations    - List Calls   - Create a Call   - Get Call by Id   - List Supervisor Inbound Calls   - Generate Call Reports  ### Phone Number Operations    - List Phone Numbers  ### Webhook Operations    - List Webhooks   - Create a Webhook   - Get Webhook by Id   - Delete Webhook by Id   - Update Webhook by Id  ### Script Operations    - List Scripts   - Create a Script   - Get Script by Id   - Delete Script by Id   - Update Script by Id  ### Voice Operations   - List Available Voices  ## Authentication🔐 Callchimp public API offers authentication with API Keys. All endpoints accepts a header named `x-api-key`. To get started follow the below steps:    1. Login to callchimp dashboard.   2. Click on your profile on the top-right corner.   3. Click on Settings.   4. On the settings page, click on `API Keys` tab.   5. Click on `Create` button, add a name and an expiry date and click on `Add`.   6. An API Key will be generated, please save the key somewhere safe as it won\'t be shown again!   7. You can use the API Key directly here in this playground to test out the APIs. 
 *
 * The version of the OpenAPI document: 1.0.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


import * as runtime from '../runtime';
import type {
  CampaignListsRequest,
  ListFindRequest,
  ListsListResponse,
  ListsRequest,
  ListsResponse,
  Model4XXResponse,
} from '../models/index';
import {
    CampaignListsRequestFromJSON,
    CampaignListsRequestToJSON,
    ListFindRequestFromJSON,
    ListFindRequestToJSON,
    ListsListResponseFromJSON,
    ListsListResponseToJSON,
    ListsRequestFromJSON,
    ListsRequestToJSON,
    ListsResponseFromJSON,
    ListsResponseToJSON,
    Model4XXResponseFromJSON,
    Model4XXResponseToJSON,
} from '../models/index';

export interface ListsCreateCampaignAndListRequest {
    campaignListsRequest: CampaignListsRequest;
}

export interface ListsDeleteRequest {
    id: number;
}

export interface ListsFindRequest {
    listFindRequest: ListFindRequest;
    page?: number;
}

export interface ListsGetRequest {
    id: number;
}

export interface ListsListRequest {
    page?: number;
}

export interface ListsPostRequest {
    listsRequest: ListsRequest;
}

export interface ListsUpdateRequest {
    id: number;
    listsRequest: ListsRequest;
}

/**
 * 
 */
export class ListsApi extends runtime.BaseAPI {

    /**
     * 
     * Create Campaign with Default List
     */
    async listsCreateCampaignAndListRaw(requestParameters: ListsCreateCampaignAndListRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<ListsResponse>> {
        if (requestParameters['campaignListsRequest'] == null) {
            throw new runtime.RequiredError(
                'campaignListsRequest',
                'Required parameter "campaignListsRequest" was null or undefined when calling listsCreateCampaignAndList().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        headerParameters['Content-Type'] = 'application/json';

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/lists/create_campaign_and_list`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: CampaignListsRequestToJSON(requestParameters['campaignListsRequest']),
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => ListsResponseFromJSON(jsonValue));
    }

    /**
     * 
     * Create Campaign with Default List
     */
    async listsCreateCampaignAndList(requestParameters: ListsCreateCampaignAndListRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<ListsResponse> {
        const response = await this.listsCreateCampaignAndListRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * Delete List by Id
     */
    async listsDeleteRaw(requestParameters: ListsDeleteRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<void>> {
        if (requestParameters['id'] == null) {
            throw new runtime.RequiredError(
                'id',
                'Required parameter "id" was null or undefined when calling listsDelete().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/lists/{Id}`.replace(`{${"Id"}}`, encodeURIComponent(String(requestParameters['id']))),
            method: 'DELETE',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);

        return new runtime.VoidApiResponse(response);
    }

    /**
     * 
     * Delete List by Id
     */
    async listsDelete(requestParameters: ListsDeleteRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<void> {
        await this.listsDeleteRaw(requestParameters, initOverrides);
    }

    /**
     * 
     * Search List by Name or Id
     */
    async listsFindRaw(requestParameters: ListsFindRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<ListsResponse>> {
        if (requestParameters['listFindRequest'] == null) {
            throw new runtime.RequiredError(
                'listFindRequest',
                'Required parameter "listFindRequest" was null or undefined when calling listsFind().'
            );
        }

        const queryParameters: any = {};

        if (requestParameters['page'] != null) {
            queryParameters['page'] = requestParameters['page'];
        }

        const headerParameters: runtime.HTTPHeaders = {};

        headerParameters['Content-Type'] = 'application/json';

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/lists/find`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: ListFindRequestToJSON(requestParameters['listFindRequest']),
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => ListsResponseFromJSON(jsonValue));
    }

    /**
     * 
     * Search List by Name or Id
     */
    async listsFind(requestParameters: ListsFindRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<ListsResponse> {
        const response = await this.listsFindRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * Get List by Id
     */
    async listsGetRaw(requestParameters: ListsGetRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<ListsResponse>> {
        if (requestParameters['id'] == null) {
            throw new runtime.RequiredError(
                'id',
                'Required parameter "id" was null or undefined when calling listsGet().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/lists/{Id}`.replace(`{${"Id"}}`, encodeURIComponent(String(requestParameters['id']))),
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => ListsResponseFromJSON(jsonValue));
    }

    /**
     * 
     * Get List by Id
     */
    async listsGet(requestParameters: ListsGetRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<ListsResponse> {
        const response = await this.listsGetRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * List Lists
     */
    async listsListRaw(requestParameters: ListsListRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<ListsListResponse>> {
        const queryParameters: any = {};

        if (requestParameters['page'] != null) {
            queryParameters['page'] = requestParameters['page'];
        }

        const headerParameters: runtime.HTTPHeaders = {};

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/lists`,
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => ListsListResponseFromJSON(jsonValue));
    }

    /**
     * 
     * List Lists
     */
    async listsList(requestParameters: ListsListRequest = {}, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<ListsListResponse> {
        const response = await this.listsListRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * Create a List
     */
    async listsPostRaw(requestParameters: ListsPostRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<ListsResponse>> {
        if (requestParameters['listsRequest'] == null) {
            throw new runtime.RequiredError(
                'listsRequest',
                'Required parameter "listsRequest" was null or undefined when calling listsPost().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        headerParameters['Content-Type'] = 'application/json';

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/lists`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: ListsRequestToJSON(requestParameters['listsRequest']),
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => ListsResponseFromJSON(jsonValue));
    }

    /**
     * 
     * Create a List
     */
    async listsPost(requestParameters: ListsPostRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<ListsResponse> {
        const response = await this.listsPostRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * Update List by Id
     */
    async listsUpdateRaw(requestParameters: ListsUpdateRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<ListsResponse>> {
        if (requestParameters['id'] == null) {
            throw new runtime.RequiredError(
                'id',
                'Required parameter "id" was null or undefined when calling listsUpdate().'
            );
        }

        if (requestParameters['listsRequest'] == null) {
            throw new runtime.RequiredError(
                'listsRequest',
                'Required parameter "listsRequest" was null or undefined when calling listsUpdate().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        headerParameters['Content-Type'] = 'application/json';

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/lists/{Id}`.replace(`{${"Id"}}`, encodeURIComponent(String(requestParameters['id']))),
            method: 'PATCH',
            headers: headerParameters,
            query: queryParameters,
            body: ListsRequestToJSON(requestParameters['listsRequest']),
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => ListsResponseFromJSON(jsonValue));
    }

    /**
     * 
     * Update List by Id
     */
    async listsUpdate(requestParameters: ListsUpdateRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<ListsResponse> {
        const response = await this.listsUpdateRaw(requestParameters, initOverrides);
        return await response.value();
    }

}
