// @dynopii/callchimp/src/apis/ScriptsApi.ts
/* tslint:disable */
/* eslint-disable */
/**
 * Callchimp Public API
 * ## Introduction👋 Introducing OpenAPI spec for doing almost anything in [callchimp.ai](https://callchimp.ai). CallChimp is a Gen AI Call Center Enhancing telecommunication with GPT-driven bulk calling. It is scalable, user-friendly, and customizable. Optimized for seamless integration and usability. ## API Categories📋 The APIs are divided in 9 categories listed below:   - Campaigns   - Supervisors   - Lists   - Subscribers   - Calls   - Phone Numbers   - Webhooks   - Scripts   - Voices  ## API Operations✅ ### Campaign Operations    - List Campaigns   - Create a Campaign   - Get Campaign by Id   - Delete Campaign by Id   - Update Campaign by Id   - Add Supervisors to Campaign by Id   - Remove Supervisors from Campaign by Id   - Upload audio file to Campaign by Id   - Search Campaign by Name or Id  ### Supervisor Operations    - List Supervisors   - Create a Supervisor   - Get Supervisor by Id   - Delete Supervisor by Id   - Update Supervisor by Id   - Send OTP to Supervisor by Id   - Verify Supervisor OTP by Id  ### List Operations    - List Lists   - Create a List   - Get List by Id   - Delete List by Id   - Update List by Id   - Search List by Name or Id   - Create Campaign with Default List  ### Subscriber Operations    - List Subscribers   - Create one or more Subscriber(s)   - Get Subscriber by Id   - Delete Subscriber by Id   - Update Subscriber by Id   - Get Subscriber by Vendor Lead Code   - Delete Subscriber by Vendor Lead Code   - Update Subscriber by Vendor Lead Code  ### Call Operations    - List Calls   - Create a Call   - Get Call by Id   - List Supervisor Inbound Calls   - Generate Call Reports  ### Phone Number Operations    - List Phone Numbers  ### Webhook Operations    - List Webhooks   - Create a Webhook   - Get Webhook by Id   - Delete Webhook by Id   - Update Webhook by Id  ### Script Operations    - List Scripts   - Create a Script   - Get Script by Id   - Delete Script by Id   - Update Script by Id  ### Voice Operations   - List Available Voices  ## Authentication🔐 Callchimp public API offers authentication with API Keys. All endpoints accepts a header named `x-api-key`. To get started follow the below steps:    1. Login to callchimp dashboard.   2. Click on your profile on the top-right corner.   3. Click on Settings.   4. On the settings page, click on `API Keys` tab.   5. Click on `Create` button, add a name and an expiry date and click on `Add`.   6. An API Key will be generated, please save the key somewhere safe as it won\'t be shown again!   7. You can use the API Key directly here in this playground to test out the APIs. 
 *
 * The version of the OpenAPI document: 1.0.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


import * as runtime from '../runtime';
import type {
  Model4XXResponse,
  ScriptListResponse,
  ScriptRequest,
  ScriptResponse,
  WebhookResponse,
} from '../models/index';
import {
    Model4XXResponseFromJSON,
    Model4XXResponseToJSON,
    ScriptListResponseFromJSON,
    ScriptListResponseToJSON,
    ScriptRequestFromJSON,
    ScriptRequestToJSON,
    ScriptResponseFromJSON,
    ScriptResponseToJSON,
    WebhookResponseFromJSON,
    WebhookResponseToJSON,
} from '../models/index';

export interface ScriptsDeleteRequest {
    id: number;
}

export interface ScriptsGetRequest {
    id: number;
}

export interface ScriptsListRequest {
    page?: number;
}

export interface ScriptsPostRequest {
    scriptRequest: ScriptRequest;
}

export interface ScriptsUpdateRequest {
    id: number;
    scriptRequest: ScriptRequest;
}

/**
 * 
 */
export class ScriptsApi extends runtime.BaseAPI {

    /**
     * 
     * Delete Script by Id
     */
    async scriptsDeleteRaw(requestParameters: ScriptsDeleteRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<void>> {
        if (requestParameters['id'] == null) {
            throw new runtime.RequiredError(
                'id',
                'Required parameter "id" was null or undefined when calling scriptsDelete().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/scripts/{Id}`.replace(`{${"Id"}}`, encodeURIComponent(String(requestParameters['id']))),
            method: 'DELETE',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);

        return new runtime.VoidApiResponse(response);
    }

    /**
     * 
     * Delete Script by Id
     */
    async scriptsDelete(requestParameters: ScriptsDeleteRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<void> {
        await this.scriptsDeleteRaw(requestParameters, initOverrides);
    }

    /**
     * 
     * Get Script by Id
     */
    async scriptsGetRaw(requestParameters: ScriptsGetRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<ScriptResponse>> {
        if (requestParameters['id'] == null) {
            throw new runtime.RequiredError(
                'id',
                'Required parameter "id" was null or undefined when calling scriptsGet().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/scripts/{Id}`.replace(`{${"Id"}}`, encodeURIComponent(String(requestParameters['id']))),
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => ScriptResponseFromJSON(jsonValue));
    }

    /**
     * 
     * Get Script by Id
     */
    async scriptsGet(requestParameters: ScriptsGetRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<ScriptResponse> {
        const response = await this.scriptsGetRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * List Script
     */
    async scriptsListRaw(requestParameters: ScriptsListRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<ScriptListResponse>> {
        const queryParameters: any = {};

        if (requestParameters['page'] != null) {
            queryParameters['page'] = requestParameters['page'];
        }

        const headerParameters: runtime.HTTPHeaders = {};

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/scripts`,
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => ScriptListResponseFromJSON(jsonValue));
    }

    /**
     * 
     * List Script
     */
    async scriptsList(requestParameters: ScriptsListRequest = {}, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<ScriptListResponse> {
        const response = await this.scriptsListRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * Create a Script
     */
    async scriptsPostRaw(requestParameters: ScriptsPostRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<ScriptResponse>> {
        if (requestParameters['scriptRequest'] == null) {
            throw new runtime.RequiredError(
                'scriptRequest',
                'Required parameter "scriptRequest" was null or undefined when calling scriptsPost().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        headerParameters['Content-Type'] = 'application/json';

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/scripts`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: ScriptRequestToJSON(requestParameters['scriptRequest']),
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => ScriptResponseFromJSON(jsonValue));
    }

    /**
     * 
     * Create a Script
     */
    async scriptsPost(requestParameters: ScriptsPostRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<ScriptResponse> {
        const response = await this.scriptsPostRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * Update Script by Id
     */
    async scriptsUpdateRaw(requestParameters: ScriptsUpdateRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<WebhookResponse>> {
        if (requestParameters['id'] == null) {
            throw new runtime.RequiredError(
                'id',
                'Required parameter "id" was null or undefined when calling scriptsUpdate().'
            );
        }

        if (requestParameters['scriptRequest'] == null) {
            throw new runtime.RequiredError(
                'scriptRequest',
                'Required parameter "scriptRequest" was null or undefined when calling scriptsUpdate().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        headerParameters['Content-Type'] = 'application/json';

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/scripts/{Id}`.replace(`{${"Id"}}`, encodeURIComponent(String(requestParameters['id']))),
            method: 'PATCH',
            headers: headerParameters,
            query: queryParameters,
            body: ScriptRequestToJSON(requestParameters['scriptRequest']),
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => WebhookResponseFromJSON(jsonValue));
    }

    /**
     * 
     * Update Script by Id
     */
    async scriptsUpdate(requestParameters: ScriptsUpdateRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<WebhookResponse> {
        const response = await this.scriptsUpdateRaw(requestParameters, initOverrides);
        return await response.value();
    }

}
