// @dynopii/callchimp/src/apis/SupervisorsApi.ts
/* tslint:disable */
/* eslint-disable */
/**
 * Callchimp Public API
 * ## Introduction👋 Introducing OpenAPI spec for doing almost anything in [callchimp.ai](https://callchimp.ai). CallChimp is a Gen AI Call Center Enhancing telecommunication with GPT-driven bulk calling. It is scalable, user-friendly, and customizable. Optimized for seamless integration and usability. ## API Categories📋 The APIs are divided in 9 categories listed below:   - Campaigns   - Supervisors   - Lists   - Subscribers   - Calls   - Phone Numbers   - Webhooks   - Scripts   - Voices  ## API Operations✅ ### Campaign Operations    - List Campaigns   - Create a Campaign   - Get Campaign by Id   - Delete Campaign by Id   - Update Campaign by Id   - Add Supervisors to Campaign by Id   - Remove Supervisors from Campaign by Id   - Upload audio file to Campaign by Id   - Search Campaign by Name or Id  ### Supervisor Operations    - List Supervisors   - Create a Supervisor   - Get Supervisor by Id   - Delete Supervisor by Id   - Update Supervisor by Id   - Send OTP to Supervisor by Id   - Verify Supervisor OTP by Id  ### List Operations    - List Lists   - Create a List   - Get List by Id   - Delete List by Id   - Update List by Id   - Search List by Name or Id   - Create Campaign with Default List  ### Subscriber Operations    - List Subscribers   - Create one or more Subscriber(s)   - Get Subscriber by Id   - Delete Subscriber by Id   - Update Subscriber by Id   - Get Subscriber by Vendor Lead Code   - Delete Subscriber by Vendor Lead Code   - Update Subscriber by Vendor Lead Code  ### Call Operations    - List Calls   - Create a Call   - Get Call by Id   - List Supervisor Inbound Calls   - Generate Call Reports  ### Phone Number Operations    - List Phone Numbers  ### Webhook Operations    - List Webhooks   - Create a Webhook   - Get Webhook by Id   - Delete Webhook by Id   - Update Webhook by Id  ### Script Operations    - List Scripts   - Create a Script   - Get Script by Id   - Delete Script by Id   - Update Script by Id  ### Voice Operations   - List Available Voices  ## Authentication🔐 Callchimp public API offers authentication with API Keys. All endpoints accepts a header named `x-api-key`. To get started follow the below steps:    1. Login to callchimp dashboard.   2. Click on your profile on the top-right corner.   3. Click on Settings.   4. On the settings page, click on `API Keys` tab.   5. Click on `Create` button, add a name and an expiry date and click on `Add`.   6. An API Key will be generated, please save the key somewhere safe as it won\'t be shown again!   7. You can use the API Key directly here in this playground to test out the APIs. 
 *
 * The version of the OpenAPI document: 1.0.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


import * as runtime from '../runtime';
import type {
  Model4XXResponse,
  SupervisorListResponse,
  SupervisorRequest,
  SupervisorResponse,
  SupervisorSendOtpResponse,
  SupervisorVerifyOtpRequest,
  SupervisorVerifyOtpResponse,
} from '../models/index';
import {
    Model4XXResponseFromJSON,
    Model4XXResponseToJSON,
    SupervisorListResponseFromJSON,
    SupervisorListResponseToJSON,
    SupervisorRequestFromJSON,
    SupervisorRequestToJSON,
    SupervisorResponseFromJSON,
    SupervisorResponseToJSON,
    SupervisorSendOtpResponseFromJSON,
    SupervisorSendOtpResponseToJSON,
    SupervisorVerifyOtpRequestFromJSON,
    SupervisorVerifyOtpRequestToJSON,
    SupervisorVerifyOtpResponseFromJSON,
    SupervisorVerifyOtpResponseToJSON,
} from '../models/index';

export interface SupervisorsDeleteRequest {
    id: number;
}

export interface SupervisorsGetRequest {
    id: number;
}

export interface SupervisorsListRequest {
    page?: number;
}

export interface SupervisorsPostRequest {
    supervisorRequest: SupervisorRequest;
}

export interface SupervisorsSendotpRequest {
    id: number;
}

export interface SupervisorsUpdateRequest {
    id: number;
    supervisorRequest: SupervisorRequest;
}

export interface SupervisorsVerifyotpRequest {
    id: number;
    supervisorVerifyOtpRequest: SupervisorVerifyOtpRequest;
}

/**
 * 
 */
export class SupervisorsApi extends runtime.BaseAPI {

    /**
     * 
     * Delete Supervisor by Id
     */
    async supervisorsDeleteRaw(requestParameters: SupervisorsDeleteRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<void>> {
        if (requestParameters['id'] == null) {
            throw new runtime.RequiredError(
                'id',
                'Required parameter "id" was null or undefined when calling supervisorsDelete().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/supervisors/{Id}`.replace(`{${"Id"}}`, encodeURIComponent(String(requestParameters['id']))),
            method: 'DELETE',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);

        return new runtime.VoidApiResponse(response);
    }

    /**
     * 
     * Delete Supervisor by Id
     */
    async supervisorsDelete(requestParameters: SupervisorsDeleteRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<void> {
        await this.supervisorsDeleteRaw(requestParameters, initOverrides);
    }

    /**
     * 
     * Get Supervisor by Id
     */
    async supervisorsGetRaw(requestParameters: SupervisorsGetRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<SupervisorResponse>> {
        if (requestParameters['id'] == null) {
            throw new runtime.RequiredError(
                'id',
                'Required parameter "id" was null or undefined when calling supervisorsGet().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/supervisors/{Id}`.replace(`{${"Id"}}`, encodeURIComponent(String(requestParameters['id']))),
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => SupervisorResponseFromJSON(jsonValue));
    }

    /**
     * 
     * Get Supervisor by Id
     */
    async supervisorsGet(requestParameters: SupervisorsGetRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<SupervisorResponse> {
        const response = await this.supervisorsGetRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * List Supervisors
     */
    async supervisorsListRaw(requestParameters: SupervisorsListRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<SupervisorListResponse>> {
        const queryParameters: any = {};

        if (requestParameters['page'] != null) {
            queryParameters['page'] = requestParameters['page'];
        }

        const headerParameters: runtime.HTTPHeaders = {};

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/supervisors`,
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => SupervisorListResponseFromJSON(jsonValue));
    }

    /**
     * 
     * List Supervisors
     */
    async supervisorsList(requestParameters: SupervisorsListRequest = {}, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<SupervisorListResponse> {
        const response = await this.supervisorsListRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * Create a Supervisor
     */
    async supervisorsPostRaw(requestParameters: SupervisorsPostRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<SupervisorResponse>> {
        if (requestParameters['supervisorRequest'] == null) {
            throw new runtime.RequiredError(
                'supervisorRequest',
                'Required parameter "supervisorRequest" was null or undefined when calling supervisorsPost().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        headerParameters['Content-Type'] = 'application/json';

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/supervisors`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: SupervisorRequestToJSON(requestParameters['supervisorRequest']),
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => SupervisorResponseFromJSON(jsonValue));
    }

    /**
     * 
     * Create a Supervisor
     */
    async supervisorsPost(requestParameters: SupervisorsPostRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<SupervisorResponse> {
        const response = await this.supervisorsPostRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * Send OTP to Supervisor by Id
     */
    async supervisorsSendotpRaw(requestParameters: SupervisorsSendotpRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<SupervisorSendOtpResponse>> {
        if (requestParameters['id'] == null) {
            throw new runtime.RequiredError(
                'id',
                'Required parameter "id" was null or undefined when calling supervisorsSendotp().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/supervisors/{Id}/send_otp`.replace(`{${"Id"}}`, encodeURIComponent(String(requestParameters['id']))),
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => SupervisorSendOtpResponseFromJSON(jsonValue));
    }

    /**
     * 
     * Send OTP to Supervisor by Id
     */
    async supervisorsSendotp(requestParameters: SupervisorsSendotpRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<SupervisorSendOtpResponse> {
        const response = await this.supervisorsSendotpRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * Update Supervisor by Id
     */
    async supervisorsUpdateRaw(requestParameters: SupervisorsUpdateRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<SupervisorResponse>> {
        if (requestParameters['id'] == null) {
            throw new runtime.RequiredError(
                'id',
                'Required parameter "id" was null or undefined when calling supervisorsUpdate().'
            );
        }

        if (requestParameters['supervisorRequest'] == null) {
            throw new runtime.RequiredError(
                'supervisorRequest',
                'Required parameter "supervisorRequest" was null or undefined when calling supervisorsUpdate().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        headerParameters['Content-Type'] = 'application/json';

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/supervisors/{Id}`.replace(`{${"Id"}}`, encodeURIComponent(String(requestParameters['id']))),
            method: 'PATCH',
            headers: headerParameters,
            query: queryParameters,
            body: SupervisorRequestToJSON(requestParameters['supervisorRequest']),
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => SupervisorResponseFromJSON(jsonValue));
    }

    /**
     * 
     * Update Supervisor by Id
     */
    async supervisorsUpdate(requestParameters: SupervisorsUpdateRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<SupervisorResponse> {
        const response = await this.supervisorsUpdateRaw(requestParameters, initOverrides);
        return await response.value();
    }

    /**
     * 
     * Verify Supervisor OTP by Id
     */
    async supervisorsVerifyotpRaw(requestParameters: SupervisorsVerifyotpRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<SupervisorVerifyOtpResponse>> {
        if (requestParameters['id'] == null) {
            throw new runtime.RequiredError(
                'id',
                'Required parameter "id" was null or undefined when calling supervisorsVerifyotp().'
            );
        }

        if (requestParameters['supervisorVerifyOtpRequest'] == null) {
            throw new runtime.RequiredError(
                'supervisorVerifyOtpRequest',
                'Required parameter "supervisorVerifyOtpRequest" was null or undefined when calling supervisorsVerifyotp().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        headerParameters['Content-Type'] = 'application/json';

        if (this.configuration && this.configuration.apiKey) {
            headerParameters["x-api-key"] = await this.configuration.apiKey("x-api-key"); // x-api-key authentication
        }

        const response = await this.request({
            path: `/supervisors/{Id}/verify_otp`.replace(`{${"Id"}}`, encodeURIComponent(String(requestParameters['id']))),
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: SupervisorVerifyOtpRequestToJSON(requestParameters['supervisorVerifyOtpRequest']),
        }, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => SupervisorVerifyOtpResponseFromJSON(jsonValue));
    }

    /**
     * 
     * Verify Supervisor OTP by Id
     */
    async supervisorsVerifyotp(requestParameters: SupervisorsVerifyotpRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<SupervisorVerifyOtpResponse> {
        const response = await this.supervisorsVerifyotpRaw(requestParameters, initOverrides);
        return await response.value();
    }

}
